torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+            optim = torch.optim.Adam(sharded_module.parameters(), lr=0.0001)
+        The optimizer will see only a single slice of parameters and will thus
+        allocate less memory for optimizer state, avoiding redundancy across
+        data parallel workers.
+        """
+        self.numel_padded_per_param = []
+        for p in self.params:
+            assert not hasattr(p, "_is_sharded")
+            assert p.is_floating_point()
+
+            # Sharding is done only when world_size is larger than 1.
+            p._is_sharded = self.world_size > 1  # type: ignore[attr-defined]
+            p._orig_size = p.size()  # type: ignore[attr-defined]
+
+            if not p._is_sharded:  # type: ignore[attr-defined]
+                self.numel_padded_per_param.append(0)
+                continue
+
+            # Replace p.data with the relevant shard.
+            orig_data = p.data
+            local_shard, num_padded = self._get_shard(p)
+            with torch.no_grad():
+                p.set_(local_shard)  # type: ignore[call-overload]
+            self.numel_padded_per_param.append(num_padded)
+            _free_storage(orig_data)
+        assert len(self.numel_padded_per_param) == len(self.params)
+
+    def _get_shard(self, tensor: torch.Tensor) -> Tuple[torch.Tensor, int]:
+        """Return the local shard of a full tensor."""
+        # Shard using torch.chunk to match all-gather/reduce-scatter.
+        chunks = list(torch.flatten(tensor).chunk(self.world_size))
+        if len(chunks) < (self.rank + 1):
+            chunk = chunks[0].new_empty(0)
+        else:
+            chunk = chunks[self.rank]
+        # Determine number of padding elements.
+        num_to_pad = chunks[0].numel() - chunk.numel()
+        assert num_to_pad >= 0
+
+        shard = chunk.clone()
+        if num_to_pad > 0:
+            shard = F.pad(shard, [0, num_to_pad])
+        return shard, num_to_pad
+
+    def __getattr__(self, name: str) -> Any:
+        """Forward missing attributes to wrapped module."""
+        try:
+            return super().__getattr__(name)  # defer to nn.Module's logic
+        except AttributeError:
+            return getattr(self.module, name)
+
+    def __getstate__(self) -> Dict[str, str]:
+        """Serialize the state of the current FSDP instance.
+        Some properties are not serializable (e.g., process groups, streams), so
+        we remove them and try to reconstruct them in :func:`__setstate__`.
+        """
+        state = copy.copy(self.__dict__)
+        state["is_sharded"] = [p._is_sharded for p in self.params]  # type: ignore[attr-defined]
+        state["orig_sizes"] = [p._orig_size for p in self.params]  # type: ignore[attr-defined]
+        if state["process_group"] is not None:
+            state["process_group"] = "MISSING"  # process_group isn't pickleable
+        self._reset_lazy_init()
+        return state
+
+    def __setstate__(self, state: Dict[str, Any]) -> None:
```

Comment:
Do you actually expect serialization of these Modules to work? Is this needed given that it is not tested?

Reply:
let me remove them until we test it